Add ansible-galaxy uninstall to remove collections and roles

s-hertel opened the pull request:
##### SUMMARY
Related, but only partially addresses #67759

Roles and collections can be uninstalled at the same time by providing a requirements file

Only direct requests are removed. Dependency removal is not supported.

Unlike legacy role removal:
- user confirmation is supported
- if a role version is specified, only a matching version will be removed
- roles without metadata can be removed
- all matching roles can be removed, instead of only the first found

##### ISSUE TYPE

- Feature Pull Request

s-hertel committed:
Add ansible-galaxy uninstall to remove collections and roles

Roles and collections can be uninstalled at the same time by providing a
requirements file

Only direct requests are removed. Dependency removal is not supported.

Unlike legacy role removal:
- user confirmation is supported
- if a role version is specified, only a matching version will be
  removed
- roles without metadata can be removed
- all matching roles can be removed, instead of only the first found

Akasurde commented on lib/ansible/galaxy/__init__.py:
```diff
@@ -22,16 +22,56 @@
+    prompt = f"Uninstalling {'collection' if collection else 'role'} {label}:"
+    prompt += (
+        "\n  Would remove:\n    " +
+        "\n    ".join(paths)
```
Add a note saying this operation is destructive and we can not recover deleted content.

s-hertel replied:
Pip doesn't provide such a disclaimer, can it recover content?

```
$ pip uninstall botocore
Found existing installation: botocore 1.36.12
Uninstalling botocore-1.36.12:
  Would remove:
    /home/shertel/.local/lib/python3.11/site-packages/botocore-1.36.12.dist-info/*
    /home/shertel/.local/lib/python3.11/site-packages/botocore/*
Proceed (Y/n)? 
```

webknjaz replied:
No, pip can't recover removed files. But I've seen some other programs in the wild that have the warning spelled out.

I checked https://clig.dev, and it doesn't seem to make an explicit recommendation. Perhaps, “Saying just enough” is applicable here, though.

s-hertel committed:
fix typo, missing return

Akasurde commented on lib/ansible/galaxy/__init__.py:
```diff
@@ -22,16 +22,56 @@
+            rc = 1
+        else:
+            display.vvvv(f"Removed {path}")
+    if rc == 0:
+        display.display(f"Uninstalled {'collection' if collection else 'role'} {label}")
```
I would say moving this block inside the for loop. This would make it easier for users to understand which collection was removed. 

If `path` is included in message then it would be more helpful in case of multiple collection paths.

s-hertel replied:
The path is displayed first in the list of paths to remove.

```
Using /home/shertel/ansible/ansible.cfg as config file
Found 1 collections to remove.
Uninstalling collection amazon.aws:0.1.1:
  Would remove:
    /home/shertel/.ansible/collections/ansible_collections/amazon/aws
    /home/shertel/.ansible/collections/ansible_collections/amazon
    /home/shertel/.ansible/collections/ansible_collections/amazon.aws-0.1.1.info

```

webknjaz replied:
What about non-interactive mode? `_ask_uninstall()` bails early and doesn't print the paths in the case of `--yes`.

s-hertel committed:
add warning for previously uninstalled requirements, similar to pip

report number of matching collections/roles found

plugin loader already reports existing collections with -vvv

s-hertel committed:
add ansible-galaxy uninstall description, fix --help

webknjaz commented:
/azp run

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -956,6 +957,54 @@ def verify_collections(
+    """Uninstall collections in any of the search paths that match a requirement."""
+    collections = []
+    for requirement in requirements:
+        namespace, name = requirement.fqcn.split(".")
+        for installed in find_existing_collections(
+            search_paths,
+            artifacts_manager,
+            namespace_filter=namespace,
+            collection_filter=name,
```
Why not access the properties directly?
```suggestion
        for installed in find_existing_collections(
            search_paths,
            artifacts_manager,
            namespace_filter=requirement.namespace,
            collection_filter=requirement.name,
```
